dom.go

```diff
@@ -709,29 +712,110 @@ func Text(text string, m ...MarkupOrChild) *HTML {
+// the component's Render method. For example, two calls to Rerender may
+// result in only one call to the component's Render method.
 func Rerender(c Component) {
 	if c == nil {
 		panic("vecty: Rerender illegally called with a nil Component argument")
 	}
-	prevRender := c.Context().prevRender
-	if prevRender == nil {
-		panic("vecty: Rerender invoked on Component that has never been rendered")
-	}
 	if c.Context().unmounted {
 		return
 	}
-	nextRender, skip, pendingMounts := renderComponent(c, prevRender)
-	if skip {
+	batch.add(c)
+}
+
+// batchRenderer handles component re-renders by queueing and deduplicating
+// them, to be rendered on the next animation frame (via requestAnimationFrame).
+type batchRenderer struct {
+	// batch contains the list of pending components to render.
+	batch []Component
+	// idx maps components to batch indexes to allow dedup, retaining order.
+	idx map[Component]int
+	// scheduled tracks whether a batch has been scheduled for processing.
+	scheduled bool
+}
+
+// add a Component to the pending batch.
+func (b *batchRenderer) add(c Component) {
+	if i, ok := b.idx[c]; ok {
+		// Delete previously queued render.
+		copy(b.batch[i:], b.batch[i+1:])
+		b.batch[len(b.batch)-1] = nil
+		b.batch = b.batch[:len(b.batch)-1]
+	}
+	// Append and index component.
+	b.batch = append(b.batch, c)
+	b.idx[c] = len(b.batch) - 1
+	// If we're not already scheduled for a render batch, request a render on
+	// the next frame.
+	if !b.scheduled {
+		b.scheduled = true
+		requestAnimationFrame(b.render)
+	}
+}
+
+// render the pending batch.
+// TODO(pdf): Add tests for time budget and multi-pass renders.
+func (b *batchRenderer) render(startTime float64) {
+	// If the batch is empty, mark as unscheduled, and stop render cycle.
+	if len(b.batch) == 0 {
+		b.scheduled = false
 		return
 	}
-	// mountUnmount to check for node replacement
-	mountUnmount(nextRender, prevRender)
-	replaceNode(nextRender.node, prevRender.node)
-	mount(pendingMounts...)
+
+	// Drain the current batch.
+	pending := b.batch
+	b.batch = nil
+	b.idx = make(map[Component]int)
+
+	// Process batch.
+	for i, c := range pending {
+		// Skip unmounted components.
+		if c.Context().unmounted {
+			continue
+		}
+
+		// Check for remaining time budget, targeting 60fps (~16ms per frame).
+		if i > 0 {
+			elapsed := global.Get("performance").Call("now").Float() - startTime
+			budgetRemaining := 16 - elapsed
+			avgRenderTime := elapsed / float64(i)
+			// If the budget remaining is less than 2 times the average
+			// Component render time, push the remainder of the batch to the
+			// next frame.
+			if budgetRemaining < avgRenderTime*2 {
+				b.batch = pending[i:]
+				for i, c := range b.batch {
+					b.idx[c] = i
+				}
+				break
+			}
+		}
+
+		// Perform render.
+		prevRender := c.Context().prevRender
+		if prevRender == nil {
+			panic("vecty: Rerender invoked on Component that has never been rendered")
+		}
```

Comment:
It just occurred to me, we should leave this check in `Rerender` since we want the panic stack trace to point to user code not to this function which gets called by the browser callback.